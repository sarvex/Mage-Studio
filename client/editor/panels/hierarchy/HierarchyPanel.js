import React from "react";
import { Tree, Input } from "antd";
import classnames from "classnames";
import {
    GatewayOutlined,
    VideoCameraOutlined,
    FolderViewOutlined,
    BulbOutlined,
} from "@ant-design/icons";

import style from "./hierarchypanel.module.scss";
import { connect } from "react-redux";
import { selectionChange } from "../../../actions/selection";

const { TreeNode } = Tree;

const ENTITY_TYPES = {
    SCENE: "SCENE",
    CAMERA: "CAMERA",
    MESH: "MESH",
    LABEL: "LABEL",
    LIGHT: {
        DEFAULT: "LIGHT.DEFAULT",
        AMBIENT: "LIGHT.AMBIENT",
        SUN: "LIGHT.SUN",
        HEMISPHERE: "LIGHT.HEMISPHERE",
        POINT: "LIGHT.POINT",
        SPOT: "LIGHT.SPOT",
    },
    AUDIO: {
        DEFAULT: "AUDIO.DEFAULT",
        DIRECTIONAL: "AUDIO.DIRECTIONAL",
        AMBIENT: "AUDIO.AMBIENT",
    },
    MODEL: "MODEL",
    SPRITE: "SPRITE",
    PARTICLE: "PARTICLE",
    EFFECT: {
        PARTICLE: "EFFECT.PARTICLE",
        SCENERY: "EFFECT.SCENERY",
    },
    HELPER: {
        GRID: "HELPER.GRID",
        AXES: "HELPER.AXES",
    },
    UNKNOWN: "UNKNOWN",
};

const DEFAULT_EXPANDED_KEY = "Level";

export class Hierarchy extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            expandedKeys: [DEFAULT_EXPANDED_KEY],
            autoExpandParent: true,
        };
    }

    onTreeNodeSelect = (_, e) => {
        const graphPosition = e.node.pos.split("-").slice(1);
        const { graph, onSelectionChange } = this.props;
        const find = (graph, indices) =>
            indices.length === 1
                ? graph[indices[0]]
                : find(graph[indices[0]].children, indices.slice(1));

        const selection = find(graph, graphPosition);

        onSelectionChange(selection);
    };

    onExpand = expandedKeys => {
        this.setState({
            expandedKeys: [...expandedKeys, DEFAULT_EXPANDED_KEY],
            autoExpandParent: false,
        });
    };

    getIcon(entityType) {
        const EntityIconComponent =
            {
                [ENTITY_TYPES.MESH]: GatewayOutlined,
                [ENTITY_TYPES.CAMERA]: VideoCameraOutlined,
                [ENTITY_TYPES.LIGHT.AMBIENT]: BulbOutlined,
                [ENTITY_TYPES.LIGHT.DEFAULT]: BulbOutlined,
                [ENTITY_TYPES.LIGHT.HEMISPHERE]: BulbOutlined,
                [ENTITY_TYPES.LIGHT.POINT]: BulbOutlined,
                [ENTITY_TYPES.LIGHT.SPOT]: BulbOutlined,
                [ENTITY_TYPES.LIGHT.SUN]: BulbOutlined,
            }[entityType] || FolderViewOutlined;

        return <EntityIconComponent height="4px" width="4px" className={style["label-icon"]} />;
    }

    render() {
        const { expandedKeys, autoExpandParent } = this.state;
        const { graph = [], selection } = this.props;
        const loop = data => {
            return data.map(({ element, children }) => {
                const name = element.name || "Level";
                if (children) {
                    return (
                        <TreeNode
                            icon={this.getIcon(element.entityType)}
                            key={element.uuid}
                            title={name}
                        >
                            {loop(children)}
                        </TreeNode>
                    );
                }
                return (
                    <TreeNode
                        icon={this.getIcon(element.entityType)}
                        key={element.uuid}
                        title={name}
                    />
                );
            });
        };

        const selectedKeys = selection.element.uuid ? [selection.element.uuid] : [];

        return (
            <div className={classnames(style.panel, style.hierarchy)}>
                <h3 className={style.title}>Hierarchy</h3>
                <div className={style.content}>
                    <Tree
                        className={style.hierarchy}
                        showIcon
                        showLine
                        onExpand={this.onExpand}
                        onSelect={this.onTreeNodeSelect}
                        expandedKeys={expandedKeys}
                        selectedKeys={selectedKeys}
                        autoExpandParent={autoExpandParent}
                    >
                        {loop(graph)}
                    </Tree>
                </div>
            </div>
        );
    }
}

const mapStateToProps = state => ({
    graph: state.hierarchy.graph,
    selection: state.selection,
});

const mapDispatchToProps = dispatch => ({
    onSelectionChange: selection => dispatch(selectionChange(selection)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Hierarchy);
