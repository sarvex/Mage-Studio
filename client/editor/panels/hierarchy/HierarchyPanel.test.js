import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import { Hierarchy } from "./HierarchyPanel";

const MOCK_GRAPH = [
    {
        element: { name: "Level" },
        children: [
            {
                element: { name: "Cube" },
                children: [],
            },
        ],
    },
];

describe("HierarchyPanel", () => {
    describe("render", () => {
        it("should render properly when empty", () => {
            const component = shallow(<Hierarchy />);
            expect(toJSON(component)).toMatchSnapshot();
        });

        it("should render properly when receives a graph", () => {
            const component = shallow(<Hierarchy graph={MOCK_GRAPH} />);
            expect(toJSON(component)).toMatchSnapshot();
        });
    });
});
