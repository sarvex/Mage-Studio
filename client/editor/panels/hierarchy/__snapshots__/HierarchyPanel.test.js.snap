// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HierarchyPanel render should render properly when empty 1`] = `
<div
  className="panel hierarchy"
>
  <h3
    className="title"
  >
    Hierarchy
  </h3>
  <div
    className="content"
  >
    <ForwardRef
      autoExpandParent={true}
      blockNode={false}
      checkable={false}
      className="hierarchy"
      expandedKeys={
        Array [
          "Level",
        ]
      }
      motion={
        Object {
          "motionAppear": false,
          "motionDeadline": 500,
          "motionName": "ant-motion-collapse",
          "onAppearActive": [Function],
          "onAppearEnd": [Function],
          "onAppearStart": [Function],
          "onEnterActive": [Function],
          "onEnterEnd": [Function],
          "onEnterStart": [Function],
          "onLeaveActive": [Function],
          "onLeaveEnd": [Function],
          "onLeaveStart": [Function],
        }
      }
      onExpand={[Function]}
      selectable={true}
      showIcon={true}
      showLine={true}
    />
  </div>
</div>
`;

exports[`HierarchyPanel render should render properly when receives a graph 1`] = `
<div
  className="panel hierarchy"
>
  <h3
    className="title"
  >
    Hierarchy
  </h3>
  <div
    className="content"
  >
    <ForwardRef
      autoExpandParent={true}
      blockNode={false}
      checkable={false}
      className="hierarchy"
      expandedKeys={
        Array [
          "Level",
        ]
      }
      motion={
        Object {
          "motionAppear": false,
          "motionDeadline": 500,
          "motionName": "ant-motion-collapse",
          "onAppearActive": [Function],
          "onAppearEnd": [Function],
          "onAppearStart": [Function],
          "onEnterActive": [Function],
          "onEnterEnd": [Function],
          "onEnterStart": [Function],
          "onLeaveActive": [Function],
          "onLeaveEnd": [Function],
          "onLeaveStart": [Function],
        }
      }
      onExpand={[Function]}
      selectable={true}
      showIcon={true}
      showLine={true}
    >
      <TreeNode
        icon={
          <ForwardRef(FolderViewOutlined)
            className="label-icon"
            height="4px"
            width="4px"
          />
        }
        key="Level"
        title="Level"
      >
        <TreeNode
          icon={
            <ForwardRef(FolderViewOutlined)
              className="label-icon"
              height="4px"
              width="4px"
            />
          }
          key="Cube"
          title="Cube"
        />
      </TreeNode>
    </ForwardRef>
  </div>
</div>
`;
